### PR TITLE
Core/PathFinding: calculate pathing with 0.5f steps and convert them to blizzard splines later on.

### LINKED ISSUE #24018: Core/Mmaps: Wrong generated path
**Description:**
If you check the npc=16708 you will see at some point of his waypoint he well go back then forward without a reason, this is caused by a wrong generated path from the mmaps, not sure if it caused by a structure above.
Check sec=16 https://mega.nz/#!1oQQxIhQ!bYPEWuaPtVETk5kqoz2B6fFhmjPPdBNGilf5AZgR7Po
Extra point shown by mmap path: https://imgur.com/a/Bw9ysbx

With mmap.enablePathFinding = 0 this issue doesn't happen

If this is a specific case I can just move the point slightly aside to fix the issue.

**Steps to reproduce the problem:**

1. .go c id 16708 
2. you can use this to remove the pause delay : UPDATE smart_scripts SET action_param1=2000 WHERE `entryorguid` IN (16708,1670800,1670801) AND action_type=54;

**Branch(es):** 3.3.5

**TC rev. hash/commit:** rev. 88a14251e276

### PR DESCRIPTION
**Pull request by Ovahlord**
**Changes proposed:**

-  use recastnavigation's default point distance for generating path points and convert them to blizzard splines after that. This will result in much more precise triangle selections to get rid of janky movement, especially when moving on stairs and other weird geometrical objects

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Issues addressed:**
closes #24018

**Tests performed:**
- tested on 4.x

### CODE REVIEW

**Comment by jackpoz** [comment]
there are a few build errors

File: src/server/game/Movement/PathGenerator.h [hunk 1]
```diff
@@ -66,7 +69,7 @@ class TC_GAME_API PathGenerator
 
         // option setters - use optional
         void SetUseStraightPath(bool useStraightPath) { _useStraightPath = useStraightPath; }
-        void SetPathLengthLimit(float distance) { _pointPathLimit = std::min<uint32>(uint32(distance/SMOOTH_PATH_STEP_SIZE), MAX_POINT_PATH_LENGTH); }
+        void SetPathLengthLimit(float distance) { _pointPathLimit = uint32(ceilf(distance / RECAST_PATH_STEP_SIZE)); }
```
Comment (jackpoz) [inline src/server/game/Movement/PathGenerator.h:72]: what happened to the std::min btw ?

Comment (Ovahlord) [reply src/server/game/Movement/PathGenerator.h:72]: not needed anymore

**Comment by Aokromes** [comment]
what's missing here?

**Comment by jackpoz** [comment]
tests or a will to merge it

**Comment by jackpoz** [comment]
@Ovahlord is this ready to be merged ?

**Comment by Jildor** [comment]
seems this can broke some things, I tested onyxia by example and seems she doesn't move in 2 phase (fly)

**Comment by Ovahlord** [comment]
Tested on 4.x and she moved just fine, so probably not related to this issue but more a general bad script as to be expected from old wotlk raids

**Comment by Jildor** [comment]
Without this PR she moves as always in phase 2, the problem is caused by some reason with this PR, I don't know what impact it can have but if things break, be old or not, it should not merge anything until it ensures that nothing breaks

**Comment by illfated** [comment]
We have seen things that work fine in 4.x not working at all in the 3.3.5 branch (Monte Muzzleshot is one of the examples). Unfortunately, we don't know if it happens due to the game client or something in the source (or a combination thereof).

**Comment by Jinnaix** [comment]
> Without this PR she moves as always in phase 2, the problem is caused by some reason with this PR, I don't know what impact it can have but if things break, be old or not, it should not merge anything until it ensures that nothing breaks

Normally I agree with you on most things but we can't just blockade new techniques/features or enhancements just because one or two old scripts start to behave weird.
This pr is a step forward and maybe we should look on how to correct older scripts in this case.

**Comment by Jildor** [comment]
> > Without this PR she moves as always in phase 2, the problem is caused by some reason with this PR, I don't know what impact it can have but if things break, be old or not, it should not merge anything until it ensures that nothing breaks
> 
> Normally I agree with you on most things but we can't just blockade new techniques/features or enhancements just because one or two old scripts start to behave weird.
> This pr is a step forward and maybe we should look on how to correct older scripts in this case.

As I said, I don't know what the impact is with this PR, I don't know if 2 scripts are broken or there are hundreds
But in case something breaks, should check why, how it should be resolved and try to repair / update it before applying anything.

**Comment by jackpoz** [comment]
It makes sense to test that boss fight with this pr and see why the boss is not moving anymore

**Comment by Aokromes** [comment]
also sindragosa, and the drake boss on hellfire and nefarian.

**Comment by jackpoz** [comment]
@Jildor could you please help me reproduce the Onyxia issue ? do you have a .go xyz ? how do I trigger Phase 2 ?

**Comment by Jildor** [comment]
Sure, go to boss onyxia (.go boss onyxia), damage the boss 40% hp and wait until boss fly, after that, boss doesn't move for the room with this PR

**Comment by funjoker** [comment]
Maybe someone can revive this PR? :)

**Comment by Jinnaix** [comment]
> seems this can broke some things, I tested onyxia by example and seems she doesn't move in 2 phase (fly)

I can **NOT**  replicate this, at least not related to changes coming from this PR.
Did several test on my fork and compared them with the changes provided here.

She was stucked twice in her flight phase without the PR on my fork.
She is borky as always otherwise on my fork and on tcubuntu without the PR.

Second testrun now with the PR changes applied, she is happily flying around.
image

Conclusion - her script can bug out. But this is not related to the PR.
I approve the changes here.

**Comment by Luzifix** [comment]
Is this change work with 9.x ?

**Comment by Jildor** [comment]
> > seems this can broke some things, I tested onyxia by example and seems she doesn't move in 2 phase (fly)
> 
> I can **NOT** replicate this, at least not related to changes coming from this PR.
> Did several test on my fork and compared them with the changes provided here.
> 
> She was stucked twice in her flight phase without the PR on my fork.
> She is borky as always otherwise on my fork and on tcubuntu without the PR.
> 
> Second testrun now with the PR changes applied, she is happily flying around.
> image
> 
> Conclusion - her script can bug out. But this is not related to the PR.
> I approve the changes here.

I can't reproduce it now, it's been too long, so I cannot indicate what possible change has been the one that has made it no longer interfere with the second phase movement (I'm not talking about the weird flight animation that she does sometimes), maybe some mmaps changes, I don't know.

**Comment by CraftedRO** [comment]
d0fd8e683ca0  I tested today this PR and Onyxia works as intended , not found any problem.
So let's port it this year il say.


https://user-images.githubusercontent.com/24683355/141281749-929b5df3-014f-4836-997e-c5c4bca17d6c.mp4

https://user-images.githubusercontent.com/24683355/141281722-4d916995-3ec6-496c-97e9-dc0ba58997d6.mp4

**Comment by Jinnaix** [comment]
@Ovahlord 

you reverted the code in this PR on your Fork without any context. Did it cause any problems or bugs that you discovered?
Can this PR be closed too?

image

**Comment by jackpoz** [comment]
I wonder if, instead of using 4.0f or 0.5f, we could instead do an hybrid solution, with the first 10/20 points with 0.5 and the rest with 4.0 . Most of the times the target will move anyway and a new path will have to be generated. This avoids the case of generating 592 points that will be discarded in the next second.

**Comment by Shauren** [comment]
Reverted in source fork, no answer here

Closing as dead

**Pull request closed without merging**